eacl/target.go

```diff
@@ -1,200 +1,174 @@
-func (t *Target) SetRole(r Role) {
-	t.role = r
-}
+	for i := range t.roles {
+		switch t.roles[i] {
+		default:
+			panic(fmt.Sprintf("invalid role #%d: forbidden value %v of enum %T", i, t.roles[i], t.roles[i]))
+		case
+			RoleContainerOwner,
+			RoleOthers:
+		}
+	}
 
-// Role returns target subject's role class.
-func (t Target) Role() Role {
-	return t.role
+	return ""
 }
 
-// ToV2 converts Target to v2 acl.EACLRecord.Target message.
-//
-// Nil Target converts to nil.
-func (t *Target) ToV2() *v2acl.Target {
-	if t == nil {
-		return nil
+// copyTo writes deep copy of the [Target] to dst.
+func (t Target) copyTo(dst *Target) {
+	if t.roles != nil {
+		dst.roles = make([]Role, len(t.roles))
+		copy(dst.roles, t.roles)
+	} else {
+		dst.roles = nil
 	}
 
-	target := new(v2acl.Target)
-	target.SetRole(t.role.ToV2())
-	target.SetKeys(t.keys)
-
-	return target
+	if t.keys != nil {
+		dst.keys = make([][]byte, len(t.keys))
+		for i := range t.keys {
+			dst.keys[i] = slice.Copy(t.keys[i])
+		}
+	} else {
+		dst.keys = nil
+	}
 }
 
-// NewTarget creates, initializes and returns blank Target instance.
+// NewTargetWithRole returns Target for the given role only.
 //
-// Defaults:
-//   - role: RoleUnknown;
-//   - keys: nil.
-func NewTarget() *Target {
-	return NewTargetFromV2(new(v2acl.Target))
+// See also [NewTarget].
+func NewTargetWithRole(role Role) Target {
+	return NewTarget([]Role{role}, nil)
 }
 
-// NewTargetFromV2 converts v2 acl.EACLRecord.Target message to Target.
-func NewTargetFromV2(target *v2acl.Target) *Target {
-	if target == nil {
-		return new(Target)
-	}
-
-	return &Target{
-		role: RoleFromV2(target.GetRole()),
-		keys: target.GetKeys(),
-	}
+// NewTargetWithKey returns Target for the given public key only.
+//
+// See also [NewTarget], [NewTargetWithKeys].
+func NewTargetWithKey(key neofscrypto.PublicKey) Target {
+	return NewTargetWithKeys([]neofscrypto.PublicKey{key})
 }
 
-// Marshal marshals Target into a protobuf binary form.
-func (t *Target) Marshal() ([]byte, error) {
-	return t.ToV2().StableMarshal(nil), nil
+// NewTargetWithKeys returns Target for the given list of public keys.
+//
+// See also [NewTarget], [NewTargetWithKey].
+func NewTargetWithKeys(publicKeys []neofscrypto.PublicKey) Target {
+	return NewTarget(nil, publicKeys)
 }
 
-// Unmarshal unmarshals protobuf binary representation of Target.
-func (t *Target) Unmarshal(data []byte) error {
-	fV2 := new(v2acl.Target)
-	if err := fV2.Unmarshal(data); err != nil {
-		return err
+// readFromV2 reads Target from the [v2acl.Target] messages. Returns an error if
+// any message is malformed according to the NeoFS API V2 protocol. Behavior is
+// forward-compatible:
+//   - unknown enum values are considered valid
+//   - unknown format of binary public keys is considered valid
+//
+// The argument MUST be non-empty.
+func (t *Target) readFromV2(ms []v2acl.Target) error {
+	if len(ms) == 0 {
+		panic("empty slice of targets")
```

Comment:
Panic? may be just an error like [here](https://github.com/nspcc-dev/neofs-sdk-go/pull/526/commits/17bfe148958ccc7b03acda65f53fdd038fcdd42e#diff-7464879ee44471d9d8b4aeacd8bee60829915e7783dbd2dc7dc2eed532c01a25R109)? We already have `error` returning value